lib/1.4/dml-builtins.dml

```diff
@@ -1058,19 +1067,55 @@ template port is (object, _qname) {
     param limitations = undefined;
     param obj = (simics_api_version == "5" || simics_api_version == "4.8")
         #? dev.obj #: _port_obj();
+    param _is_subdevice_port : bool;
+    param _is_subdevice_port default false;
+    #if (this._nongroup_parent.objtype == "port"
+         #? !this._nongroup_parent._is_subdevice_port #: false) {
+        error "Nested ports are only permitted if the outer port instantiates"
+            + " the subdevice_port template";
+    }
+
     session conf_object_t *_cached_port_obj;
     shared method _port_obj() -> (conf_object_t *) {
         if (this._cached_port_obj != NULL) {
             return this._cached_port_obj;
         }
-        local char name[strlen(this._qname()) + 6];
-        strcpy(name, "port.");
-        strcpy(name + 5, this._qname());
+        local const char *qn = this._qname();
+        local int len = strlen(qn);
+        local int anc_len;
+        for (anc_len = strlen(qn); anc_len > 0; --anc_len) {
+            if (qn[anc_len - 1] == '.')
+                break;
+        }
+        local char name[len + 6];
+        memcpy(name, qn, anc_len);
+        memcpy(name + anc_len, "port.", 5);
+        strcpy(name + anc_len + 5, qn + anc_len);
```

Comment:
Why not `snprintf`?
```
        local char name[len + 6];
        memcpy(name, qn, anc_len);
        snprintf(name + anc_len, len + 6 - anc_len, "port.%s", qn + anc_len);
```

Reply:
Because it's not available in DML scope, and adding/removing wrapped symbols in dml-builtins.dml comes with a cost (currently, if a user needs snprintf, she must write `extern int snprintf(...);`, and if we add a symbol then existing such uses will ENAMECOLL). This is also the reason for the locally inlined `strrchr`.

Reply:
Oh right, that classic issue. Still think we should look into making `extern` declarations idempotent.

Reply:
We should, but I think the type system needs to grow just a bit more robust first. Must be able to trust type equivelence.